File: .samples/javascript/irritating_parameter/bankconnector.ts
export type OAuthTokenType = string;
export const OAuthToken = 'A really complicated OAuth token that requires calling an external API';

// After start
export abstract class TokenProvider {
  abstract token(): OAuthTokenType;
}

class OAuthTokenProvider extends TokenProvider {
  token(): OAuthTokenType {
    return OAuthToken;
  }
}
// After end

export class BankConnector {

  // After start
  static create(baseUrl: string, provider: TokenProvider): BankConnector {
    return new BankConnector(baseUrl, provider.token());
  }
  // After end

  // Uncomment:  constructor(public baseUrl: string, public token: OAuthTokenType) {
  // After start
  constructor(public baseUrl: string, public token: OAuthTokenType | undefined) {
  // After end
    this.baseUrl = baseUrl;
    this.token = token;
  }

  _buildURL(start: Date, end: Date): string {
    const sstart = start.toISOString().slice(0, 10);
    const send = end.toISOString().slice(0, 10);
    return `${this.baseUrl}rest/merchant/${sstart}/${send}/transactions.csv`;
  }

  getPayments(start: Date, end: Date): string {
    const url = this._buildURL(start, end);
    const headers = { Authorization: `Token ${this.token}` };
    return this.getResponse(url, headers);
  }

  getResponse(url: string, headers: Record<string, string>): string {
    // does nothing in this exercise
    return headers['Authorization'];
  }
}

File: .samples/python/irritating_parameter/bankconnector.py

from abc import ABC, abstractmethod
from datetime import datetime

OAuthTokenType = str
OAuthToken = "A really complicated OAuth token that requires calling an external API"

# After start
class TokenProvider(ABC):
    @abstractmethod
    def token(self) -> OAuthTokenType:
        pass

class OAuthTokenProvider(TokenProvider):
    def token(self) -> OAuthTokenType:
        return OAuthToken
# After end

class BankConnector:
    # After start
    @classmethod
    def create(cls, baseUrl: str, provider: TokenProvider):
        return cls(baseUrl, provider.token)
    # After end
    
    def __init__(self, baseUrl: str, token: OAuthTokenType):
        self.baseUrl = baseUrl
        self.token = token
        
    def _buildURL(self, start: datetime, end: datetime) -> str:
        sstart = start.strftime("%Y-%m-%d")
        send = end.strftime("%Y-%m-%d")
        return f"{self.baseUrl}rest/merchant/{sstart}/{send}/transactions.csv"
    
    def getPayments(self,start: datetime,end:datetime):
        url = self._buildURL(start, end)
        headers = {"Authorization": f"Token {self.token}"}
        return self.getResponse(url, headers)
    
    def getResponse(self, url: str, headers: dict):
        # does nothing in this exercise
        return headers["Authorization"]
        